File: golang/task1/handle-function/handle-api-query.go
package handle_function

import (
	"encoding/json"
	"github.com/tealeg/xlsx"
	"net/http"
	"strconv"
	"strings"
)

func HandleApiQuery(w http.ResponseWriter, r *http.Request, uploadedFile *xlsx.File) {
	if uploadedFile == nil {
		http.Error(w, "Vui long tai len file truoc khi truy van", http.StatusBadRequest)
		return
	}

	gioBatDau := r.URL.Query().Get("gio-bat-dau")
	gioKetThuc := r.URL.Query().Get("gio-ket-thuc")

	if gioBatDau == "" || gioKetThuc == "" || len(gioBatDau) != 6 || len(gioKetThuc) != 6 {
		http.Error(w, "Sai dinh dang truy van, thu lai voi gio-bat-dau va gio-ket-thuc hhmmss", http.StatusBadRequest)
		return
	}

	startTime, err := strconv.Atoi(gioBatDau)
	if err != nil {
		http.Error(w, "Invalid start time format", http.StatusBadRequest)
		return
	}

	endTime, err := strconv.Atoi(gioKetThuc)
	if err != nil {
		http.Error(w, "Invalid end time format", http.StatusBadRequest)
		return
	}

	if startTime > endTime {
		http.Error(w, "Start time must be less than end time", http.StatusBadRequest)
		return
	}

	sheet := uploadedFile.Sheets[0]
	totalSum := 0.0

	for _, row := range sheet.Rows[8:] {
		timeCell := row.Cells[2]
		totalPriceCell := row.Cells[8]

		if timeCell != nil && totalPriceCell != nil {
			time := strings.ReplaceAll(timeCell.String(), ":", "")
			timeInt, err := strconv.Atoi(time)
			if err != nil {
				continue
			}

			if timeInt >= startTime && timeInt <= endTime {
				totalPrice, err := totalPriceCell.Float()
				if err != nil {
					continue
				}
				totalSum += totalPrice
			}
		}
	}

	result := map[string]float64{"totalSum": totalSum}
	json.NewEncoder(w).Encode(result)
}

File: node-js/task1/handle-function/handle-api-query.js
const xlsx = require("xlsx");

function handleApiQuery(req, res, uploadedFile) {
    // kiem tra xem da upload file chua
    if (!uploadedFile) {
        res.status(400).send('Vui long tai len file truoc khi truy van.');
        return;
    }
    const gio_bat_dau = req.query["gio-bat-dau"];
    const gio_ket_thuc = req.query["gio-ket-thuc"];

    // kiem tra xem gio bat dau va gio ket thuc co dung dinh dang hhmmss khong
    if (gio_bat_dau && gio_ket_thuc && /^\d{6}$/.test(gio_bat_dau) && /^\d{6}$/.test(gio_ket_thuc)) {
        if (parseInt(gio_bat_dau) > parseInt(gio_ket_thuc)) {
            res.status(400).send('Gio bat dau phai nho hon gio ket thuc.');
            return;
        }
        if (gio_bat_dau < 0 || gio_ket_thuc > 235959) {
            res.status(400).send('Gio bat dau va gio ket thuc phai nam trong khoang 000000 - 235959.');
            return;
        }
        // lay sheet dau tien
        const sheet = uploadedFile.Sheets[uploadedFile.SheetNames[0]];
        const range = xlsx.utils.decode_range(sheet['!ref']);
        let totalSum = 0;

        for (let i = range.s.r + 8; i <= range.e.r; i++) {
            // lay cell thoi gian va cell tong tien
            const timeCell = sheet[xlsx.utils.encode_cell({r: i, c: 2})];
            const totalPriceCell = sheet[xlsx.utils.encode_cell({r: i, c: 8})];

            if (timeCell && totalPriceCell) {
                // lay thoi gian va tong tien
                const time = timeCell.v;
                const totalPrice = parseFloat(totalPriceCell.v);

                if (time >= gio_bat_dau && time <= gio_ket_thuc) {
                    // neu thoi gian nam trong khoang gio bat dau va gio ket thuc thi cong tong tien vao total
                    totalSum += totalPrice;
                }
            }
        }

        res.json({totalSum});
    } else {
        res.status(400).send('Sai định dạng thời gian. Vui lòng nhập lại hhmmss');
    }
}

module.exports = handleApiQuery;